courses/fundamentals_of_ada/adv_250_containers.rst

```diff
@@ -35,69 +35,255 @@ Introduction
+    - Dynamically sized (indefinite elements)
+    - Random access in ``O(1)``
+
+* :ada:`Ordered_Set`
+
+    - Unique elements
+    - Differenciated by :ada:`<` and :ada:`=`
+    - Manipulated in order
+
+* :ada:`Bounded_Doubly_Linked_List`
+
+    - Static size of container and elements
+    - Insertions and deletions in ``O(1)``
+
+-------------
+Declaration
+-------------
+
+* Generic packages
+* Always need at least the ``Element_Type``
+* Examples chosen for the next slides:
+
+.. include:: examples/containers/extracts/decl_vector.ads
+    :code: Ada
+
+.. include:: examples/containers/extracts/decl_set.ads
+    :code: Ada
+
+.. include:: examples/containers/extracts/decl_map.ads
+    :code: Ada
+
+---------------
+Instanciation
+---------------
+
+* May require an initial :ada:`Empty_xxx` value
+
+.. include:: examples/containers/extracts/decl_instances.adb
+    :code: Ada
+
+=======================
+Containers Operations
+=======================
 
 -------------------
-Library Structure
+Common Operations
 -------------------
 
-* Groups of objects of
+* Lots of common operations
+
+    - What is available depends greatly on the exact container type
+    - ... so does syntax
+
+* Insertion
+* Iteration
+* Comparison
+* Sort
+* Search
+
+-----------
+Insertion
+-----------
+
+* May be in order :ada:`Append` or :ada:`Prepend`
+* May be :ada:`Insert` (at random or at given index)
+* May :ada:`Replace` an existing element
 
-   - Definite types
-   - Indefinite types
-   - Bounded capacity (of definite types)
+.. include:: examples/containers/extracts/insert.adb
+    :code: Ada
 
-* Definite types
+-----------
+Iteration
+-----------
 
-   - More efficient
+* Container have a :ada:`Cursor` type
 
-* Indefinite types
+    - Points to an element in a container
+    - Can be used for advanced iterations
 
-   - More flexible
+.. include:: examples/containers/extracts/iterate.adb
+    :code: Ada
 
-* Bounded capacity
+------------
+Comparison
+------------
 
-   - Known size
+.. include:: examples/containers/extracts/compare.adb
+    :code: Ada
 
------------------------------
-Common Container Processing
------------------------------
+------
+Sort
+------
 
-* Every container has a type `Cursor` which points to an element in a container
+* Arrays
+
+   - `Ada.Containers.Generic_Array_Sort`
+   - `Ada.Containers.Generic_Constrained_Array_Sort`
 
-   - Points to the container as well as the element!
+* Any object that has indexing
 
-* Container types are tagged allowing for prefixed notation
+   - `Ada.Containers.Generic_Sort`
 
-   - Prefixed: `My_List.Append ( Item );`
-   - Untagged: `Append(My_List, Item);`
+.. include:: examples/containers/extracts/sort.adb
+    :code: Ada
 
-* Hashed vs Ordered
+--------
+Search
+--------
 
-   - **Ordered** containers are sequential
+* Use :ada:`Find` for a :ada:`Cursor`
 
-      + Searching can be slow for large containers
+    - :ada:`<Pkg>.No_Element` is a :ada:`Cursor` if not found
```

Comment:
Should read more like "No_Element" is an "empty" cursor for when Find is unsuccessful